src/core/server/http/router/socket.ts

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Socket } from 'net';
+import { DetailedPeerCertificate, PeerCertificate, TLSSocket } from 'tls';
+
+/**
+ * A tiny abstraction for TCP socket.
+ * @public
+ */
+export interface IKibanaSocket {
+  getPeerCertificate(detailed: true): DetailedPeerCertificate | null;
+  getPeerCertificate(detailed: false): PeerCertificate | null;
+  /**
+   * Returns an object representing the peer's certificate.
+   * The returned object has some properties corresponding to the field of the certificate.
+   * If detailed argument is true the full chain with issuer property will be returned,
+   * if false only the top certificate without issuer property.
+   * If the peer does not provide a certificate, it returns null.
+   * @param detailed - If true; the full chain with issuer property will be returned.
+   * @returns An object representing the peer's certificate.
+   */
+  getPeerCertificate(detailed?: boolean): PeerCertificate | DetailedPeerCertificate | null;
+}
+
+export class KibanaSocket implements IKibanaSocket {
+  constructor(private readonly socket: Socket) {}
+
+  getPeerCertificate(detailed: true): DetailedPeerCertificate | null;
+  getPeerCertificate(detailed: false): PeerCertificate | null;
+  getPeerCertificate(detailed?: boolean): PeerCertificate | DetailedPeerCertificate | null;
+
+  public getPeerCertificate(detailed?: boolean) {
+    if (this.socket instanceof TLSSocket) {
```

Comment:
__note:__ PKI authentication provider will consume certificate chain returned from this method, but it works on the assumption that `rejectUnauthorized` is set to `true` in the core server and hence we don't need to verify `socket.authorized`. `rejectUnauthorized` in the core server isn't configurable right now and hence relies on Node's default value which is `true`.

So everything is fine right now, but in case we decide to expose `rejectUnauthorized` as a configuration option for some reason (hopefully not) we may need to expose `socket.authorized` from `IKibanaSocket`. Just wanted to note.

/cc @kobelb